File: packages/less/src/less/source_map_output_test.go
package less

import (
	"testing"
)

type MockSourceMapGenerator struct {
	Mappings       []SourceMapMapping
	SourceContents map[string]string
}

func (m *MockSourceMapGenerator) AddMapping(mapping SourceMapMapping) {
	m.Mappings = append(m.Mappings, mapping)
}

func (m *MockSourceMapGenerator) SetSourceContent(source, content string) {
	if m.SourceContents == nil {
		m.SourceContents = make(map[string]string)
	}
	m.SourceContents[source] = content
}

func (m *MockSourceMapGenerator) ToJSON() map[string]any {
	return map[string]any{
		"version":  3,
		"sources":  []string{"test.less"},
		"mappings": "AAAA",
	}
}

type MockNode struct {
	GenCSSFunc func(context map[string]any, output *SourceMapOutput)
}

func (m *MockNode) GenCSS(context map[string]any, output *SourceMapOutput) {
	if m.GenCSSFunc != nil {
		m.GenCSSFunc(context, output)
	}
}

func createMockSourceMapGenerator() SourceMapGenerator {
	return &MockSourceMapGenerator{}
}

func TestSourceMapOutput(t *testing.T) {
	t.Run("constructor", func(t *testing.T) {
		t.Run("should initialize with basic options", func(t *testing.T) {
			mockRootNode := &MockNode{}
			options := SourceMapOutputOptions{
				RootNode:                mockRootNode,
				ContentsMap:            map[string]string{"test.less": "body { color: red; }"},
				ContentsIgnoredCharsMap: map[string]int{},
				OutputFilename:         "output.css",
			}

			sourceMapOutput := NewSourceMapOutput(options)

			if len(sourceMapOutput.css) != 0 {
				t.Errorf("css should be empty initially, got %v", sourceMapOutput.css)
			}
			if sourceMapOutput.rootNode != mockRootNode {
				t.Errorf("rootNode should be set")
			}
			if sourceMapOutput.outputFilename != "output.css" {
				t.Errorf("outputFilename = %v, want %v", sourceMapOutput.outputFilename, "output.css")
			}
			if sourceMapOutput.lineNumber != 0 {
				t.Errorf("lineNumber = %v, want %v", sourceMapOutput.lineNumber, 0)
			}
			if sourceMapOutput.column != 0 {
				t.Errorf("column = %v, want %v", sourceMapOutput.column, 0)
			}
		})

		t.Run("should normalize Windows paths in sourceMapFilename", func(t *testing.T) {
			options := SourceMapOutputOptions{
				RootNode:               &MockNode{},
				ContentsMap:            map[string]string{},
				ContentsIgnoredCharsMap: map[string]int{},
				SourceMapFilename:      "C:\\path\\to\\map.css.map",
			}

			sourceMapOutput := NewSourceMapOutput(options)
			if sourceMapOutput.sourceMapFilename != "C:/path/to/map.css.map" {
				t.Errorf("sourceMapFilename = %v, want %v", sourceMapOutput.sourceMapFilename, "C:/path/to/map.css.map")
			}
		})

		t.Run("should normalize and add trailing slash to sourceMapRootpath", func(t *testing.T) {
			options := SourceMapOutputOptions{
				RootNode:               &MockNode{},
				ContentsMap:            map[string]string{},
				ContentsIgnoredCharsMap: map[string]int{},
				SourceMapRootpath:      "C:\\src\\styles",
			}

			sourceMapOutput := NewSourceMapOutput(options)
			if sourceMapOutput.sourceMapRootpath != "C:/src/styles/" {
				t.Errorf("sourceMapRootpath = %v, want %v", sourceMapOutput.sourceMapRootpath, "C:/src/styles/")
			}
		})

		t.Run("should handle sourceMapRootpath that already has trailing slash", func(t *testing.T) {
			options := SourceMapOutputOptions{
				RootNode:               &MockNode{},
				ContentsMap:            map[string]string{},
				ContentsIgnoredCharsMap: map[string]int{},
				SourceMapRootpath:      "/src/styles/",
			}

			sourceMapOutput := NewSourceMapOutput(options)
			if sourceMapOutput.sourceMapRootpath != "/src/styles/" {
				t.Errorf("sourceMapRootpath = %v, want %v", sourceMapOutput.sourceMapRootpath, "/src/styles/")
			}
		})

		t.Run("should set empty sourceMapRootpath when not provided", func(t *testing.T) {
			options := SourceMapOutputOptions{
				RootNode:               &MockNode{},
				ContentsMap:            map[string]string{},
				ContentsIgnoredCharsMap: map[string]int{},
			}

			sourceMapOutput := NewSourceMapOutput(options)
			if sourceMapOutput.sourceMapRootpath != "" {
				t.Errorf("sourceMapRootpath = %v, want %v", sourceMapOutput.sourceMapRootpath, "")
			}
		})

		t.Run("should normalize Windows paths in sourceMapBasepath", func(t *testing.T) {
			options := SourceMapOutputOptions{
				RootNode:               &MockNode{},
				ContentsMap:            map[string]string{},
				ContentsIgnoredCharsMap: map[string]int{},
				SourceMapBasepath:      "C:\\project\\src",
			}

			sourceMapOutput := NewSourceMapOutput(options)
			if sourceMapOutput.sourceMapBasepath != "C:/project/src" {
				t.Errorf("sourceMapBasepath = %v, want %v", sourceMapOutput.sourceMapBasepath, "C:/project/src")
			}
		})
	})

	t.Run("RemoveBasepath", func(t *testing.T) {
		options := SourceMapOutputOptions{
			RootNode:               &MockNode{},
			ContentsMap:            map[string]string{},
			ContentsIgnoredCharsMap: map[string]int{},
			SourceMapBasepath:      "/project/src",
		}
		sourceMapOutput := NewSourceMapOutput(options)

		t.Run("should remove basepath when path starts with basepath", func(t *testing.T) {
			result := sourceMapOutput.RemoveBasepath("/project/src/styles/main.less")
			if result != "styles/main.less" {
				t.Errorf("RemoveBasepath() = %v, want %v", result, "styles/main.less")
			}
		})

		t.Run("should remove basepath and leading slash/backslash", func(t *testing.T) {
			sourceMapOutput.sourceMapBasepath = "/project/src"
			if got := sourceMapOutput.RemoveBasepath("/project/src/main.less"); got != "main.less" {
				t.Errorf("RemoveBasepath() = %v, want %v", got, "main.less")
			}
			if got := sourceMapOutput.RemoveBasepath("/project/src\\main.less"); got != "main.less" {
				t.Errorf("RemoveBasepath() = %v, want %v", got, "main.less")
			}
		})

		t.Run("should return original path when basepath does not match", func(t *testing.T) {
			result := sourceMapOutput.RemoveBasepath("/other/path/main.less")
			if result != "/other/path/main.less" {
				t.Errorf("RemoveBasepath() = %v, want %v", result, "/other/path/main.less")
			}
		})

		t.Run("should handle empty basepath", func(t *testing.T) {
			sourceMapOutput.sourceMapBasepath = ""
			result := sourceMapOutput.RemoveBasepath("/path/to/file.less")
			if result != "/path/to/file.less" {
				t.Errorf("RemoveBasepath() = %v, want %v", result, "/path/to/file.less")
			}
		})
	})

	t.Run("NormalizeFilename", func(t *testing.T) {
		options := SourceMapOutputOptions{
			RootNode:               &MockNode{},
			ContentsMap:            map[string]string{},
			ContentsIgnoredCharsMap: map[string]int{},
			SourceMapBasepath:      "/project/src",
			SourceMapRootpath:      "/assets/",
		}
		sourceMapOutput := NewSourceMapOutput(options)

		t.Run("should normalize Windows paths to forward slashes", func(t *testing.T) {
			result := sourceMapOutput.NormalizeFilename("C:\\project\\src\\styles\\main.less")
			if result != "/assets/C:/project/src/styles/main.less" {
				t.Errorf("NormalizeFilename() = %v, want %v", result, "/assets/C:/project/src/styles/main.less")
			}
		})

		t.Run("should combine rootpath with filename after removing basepath", func(t *testing.T) {
			result := sourceMapOutput.NormalizeFilename("/project/src/main.less")
			if result != "/assets/main.less" {
				t.Errorf("NormalizeFilename() = %v, want %v", result, "/assets/main.less")
			}
		})

		t.Run("should handle empty rootpath", func(t *testing.T) {
			sourceMapOutput.sourceMapRootpath = ""
			result := sourceMapOutput.NormalizeFilename("/project/src/main.less")
			if result != "main.less" {
				t.Errorf("NormalizeFilename() = %v, want %v", result, "main.less")
			}
		})
	})

	t.Run("Add", func(t *testing.T) {
		options := SourceMapOutputOptions{
			RootNode: &MockNode{},
			ContentsMap: map[string]string{
				"test.less": "body {\n  color: red;\n}",
			},
			ContentsIgnoredCharsMap: map[string]int{},
		}
		sourceMapOutput := NewSourceMapOutput(options)

		t.Run("should ignore empty chunks", func(t *testing.T) {
			sourceMapOutput.Add("", nil, 0, false)
			
			if len(sourceMapOutput.css) != 0 {
				t.Errorf("css should be empty, got %v", sourceMapOutput.css)
			}
		})

		t.Run("should add chunk without file info", func(t *testing.T) {
			sourceMapOutput.Add("body { color: blue; }", nil, 0, false)
			
			expected := []string{"body { color: blue; }"}
			if len(sourceMapOutput.css) != 1 || sourceMapOutput.css[0] != expected[0] {
				t.Errorf("css = %v, want %v", sourceMapOutput.css, expected)
			}
			if sourceMapOutput.column != 21 {
				t.Errorf("column = %v, want %v", sourceMapOutput.column, 21)
			}
			if sourceMapOutput.lineNumber != 0 {
				t.Errorf("lineNumber = %v, want %v", sourceMapOutput.lineNumber, 0)
			}
		})

		t.Run("should handle multi-line chunks", func(t *testing.T) {
			sourceMapOutput := NewSourceMapOutput(options) // fresh instance
			sourceMapOutput.Add("body {\n  color: blue;\n}", nil, 0, false)
			
			if sourceMapOutput.lineNumber != 2 {
				t.Errorf("lineNumber = %v, want %v", sourceMapOutput.lineNumber, 2)
			}
			if sourceMapOutput.column != 1 {
				t.Errorf("column = %v, want %v", sourceMapOutput.column, 1)
			}
		})

		t.Run("should add chunk with file info and create mapping", func(t *testing.T) {
			sourceMapOutput := NewSourceMapOutput(options) // fresh instance
			mockGen := &MockSourceMapGenerator{}
			sourceMapOutput.sourceMapGenerator = mockGen
			fileInfo := &FileInfo{Filename: "test.less"}
			
			sourceMapOutput.Add("body { color: blue; }", fileInfo, 5, false)
			
			if len(mockGen.Mappings) != 1 {
				t.Errorf("Expected 1 mapping, got %d", len(mockGen.Mappings))
			}
			if len(mockGen.Mappings) > 0 {
				mapping := mockGen.Mappings[0]
				if mapping.Generated.Line != 1 || mapping.Generated.Column != 0 {
					t.Errorf("Generated position = %v, want {1, 0}", mapping.Generated)
				}
				if mapping.Original.Line != 1 || mapping.Original.Column != 5 {
					t.Errorf("Original position = %v, want {1, 5}", mapping.Original)
				}
				if mapping.Source != "test.less" {
					t.Errorf("Source = %v, want %v", mapping.Source, "test.less")
				}
			}
		})

		t.Run("should handle mapLines option", func(t *testing.T) {
			sourceMapOutput := NewSourceMapOutput(options) // fresh instance
			mockGen := &MockSourceMapGenerator{}
			sourceMapOutput.sourceMapGenerator = mockGen
			fileInfo := &FileInfo{Filename: "test.less"}
			
			sourceMapOutput.Add("line1\nline2\nline3", fileInfo, 0, true)
			
			if len(mockGen.Mappings) != 3 {
				t.Errorf("Expected 3 mappings, got %d", len(mockGen.Mappings))
			}
			if len(mockGen.Mappings) > 0 {
				mapping := mockGen.Mappings[0]
				if mapping.Generated.Line != 1 || mapping.Generated.Column != 0 {
					t.Errorf("First mapping generated position = %v, want {1, 0}", mapping.Generated)
				}
				if mapping.Original.Line != 1 || mapping.Original.Column != 0 {
					t.Errorf("First mapping original position = %v, want {1, 0}", mapping.Original)
				}
			}
			if len(mockGen.Mappings) > 1 {
				mapping := mockGen.Mappings[1]
				if mapping.Generated.Line != 2 || mapping.Generated.Column != 0 {
					t.Errorf("Second mapping generated position = %v, want {2, 0}", mapping.Generated)
				}
			}
		})

		t.Run("should handle ignored characters", func(t *testing.T) {
			options := SourceMapOutputOptions{
				RootNode: &MockNode{},
				ContentsMap: map[string]string{
					"test.less": "body {\n  color: red;\n}",
				},
				ContentsIgnoredCharsMap: map[string]int{"test.less": 5},
			}
			sourceMapOutput := NewSourceMapOutput(options)
			mockGen := &MockSourceMapGenerator{}
			sourceMapOutput.sourceMapGenerator = mockGen
			fileInfo := &FileInfo{Filename: "test.less"}
			
			sourceMapOutput.Add("color: blue;", fileInfo, 10, false)
			
			if len(mockGen.Mappings) != 1 {
				t.Errorf("Expected 1 mapping, got %d", len(mockGen.Mappings))
			}
			if len(mockGen.Mappings) > 0 {
				mapping := mockGen.Mappings[0]
				if mapping.Generated.Line != 1 || mapping.Generated.Column != 0 {
					t.Errorf("Generated position = %v, want {1, 0}", mapping.Generated)
				}
				if mapping.Original.Line != 2 || mapping.Original.Column != 3 {
					t.Errorf("Original position = %v, want {2, 3}", mapping.Original)
				}
				if mapping.Source != "test.less" {
					t.Errorf("Source = %v, want %v", mapping.Source, "test.less")
				}
			}
		})

		t.Run("should handle negative index after ignored chars adjustment", func(t *testing.T) {
			options := SourceMapOutputOptions{
				RootNode: &MockNode{},
				ContentsMap: map[string]string{
					"test.less": "body {\n  color: red;\n}",
				},
				ContentsIgnoredCharsMap: map[string]int{"test.less": 10},
			}
			sourceMapOutput := NewSourceMapOutput(options)
			mockGen := &MockSourceMapGenerator{}
			sourceMapOutput.sourceMapGenerator = mockGen
			fileInfo := &FileInfo{Filename: "test.less"}
			
			sourceMapOutput.Add("color: blue;", fileInfo, 5, false)
			
			if len(mockGen.Mappings) != 1 {
				t.Errorf("Expected 1 mapping, got %d", len(mockGen.Mappings))
			}
			if len(mockGen.Mappings) > 0 {
				mapping := mockGen.Mappings[0]
				if mapping.Original.Line != 1 || mapping.Original.Column != 0 {
					t.Errorf("Original position = %v, want {1, 0}", mapping.Original)
				}
			}
		})

		t.Run("should handle missing content in contentsMap", func(t *testing.T) {
			sourceMapOutput := NewSourceMapOutput(options) // fresh instance
			fileInfo := &FileInfo{Filename: "missing.less"}
			
			sourceMapOutput.Add("body { color: blue; }", fileInfo, 0, false)
			
			expected := []string{"body { color: blue; }"}
			if len(sourceMapOutput.css) != 1 || sourceMapOutput.css[0] != expected[0] {
				t.Errorf("css = %v, want %v", sourceMapOutput.css, expected)
			}
		})
	})

	t.Run("IsEmpty", func(t *testing.T) {
		options := SourceMapOutputOptions{
			RootNode:               &MockNode{},
			ContentsMap:            map[string]string{},
			ContentsIgnoredCharsMap: map[string]int{},
		}
		sourceMapOutput := NewSourceMapOutput(options)

		t.Run("should return true when no CSS has been added", func(t *testing.T) {
			if !sourceMapOutput.IsEmpty() {
				t.Error("IsEmpty() should return true for empty output")
			}
		})

		t.Run("should return false when CSS has been added", func(t *testing.T) {
			sourceMapOutput.Add("body { color: red; }", nil, 0, false)
			if sourceMapOutput.IsEmpty() {
				t.Error("IsEmpty() should return false after adding CSS")
			}
		})
	})

	t.Run("ToCSS", func(t *testing.T) {
		mockContext := map[string]any{}
		options := SourceMapOutputOptions{
			RootNode: &MockNode{},
			ContentsMap: map[string]string{
				"test.less": "body { color: red; }",
			},
			ContentsIgnoredCharsMap: map[string]int{},
			OutputFilename:         "output.css",
			OutputSourceFiles:      true,
			SourceMapGeneratorConstructor: createMockSourceMapGenerator,
		}
		sourceMapOutput := NewSourceMapOutput(options)

		t.Run("should generate CSS and source map", func(t *testing.T) {
			sourceMapOutput.rootNode = &MockNode{
				GenCSSFunc: func(context map[string]any, output *SourceMapOutput) {
					output.Add("body { color: red; }", nil, 0, false)
				},
			}

			css := sourceMapOutput.ToCSS(mockContext)
			
			if css != "body { color: red; }" {
				t.Errorf("ToCSS() = %v, want %v", css, "body { color: red; }")
			}
			if sourceMapOutput.SourceMap == "" {
				t.Error("SourceMap should be defined")
			}
		})

		t.Run("should set source content when outputSourceFiles is true", func(t *testing.T) {
			sourceMapOutput.rootNode = &MockNode{
				GenCSSFunc: func(context map[string]any, output *SourceMapOutput) {
					output.Add("body { color: red; }", nil, 0, false)
				},
			}

			sourceMapOutput.ToCSS(mockContext)
			
			mockGen, ok := sourceMapOutput.sourceMapGenerator.(*MockSourceMapGenerator)
			if !ok {
				t.Fatal("Expected MockSourceMapGenerator")
			}
			if content, exists := mockGen.SourceContents["test.less"]; !exists || content != "body { color: red; }" {
				t.Errorf("Source content not set correctly: %v", mockGen.SourceContents)
			}
		})

		t.Run("should handle ignored characters in source content", func(t *testing.T) {
			options := SourceMapOutputOptions{
				RootNode: &MockNode{},
				ContentsMap: map[string]string{
					"test.less": "/* */body { color: red; }",
				},
				ContentsIgnoredCharsMap: map[string]int{"test.less": 5},
				OutputSourceFiles:      true,
				SourceMapGeneratorConstructor: createMockSourceMapGenerator,
			}
			sourceMapOutput := NewSourceMapOutput(options)
			sourceMapOutput.rootNode = &MockNode{
				GenCSSFunc: func(context map[string]any, output *SourceMapOutput) {
					output.Add("body { color: red; }", nil, 0, false)
				},
			}

			sourceMapOutput.ToCSS(mockContext)
			
			mockGen, ok := sourceMapOutput.sourceMapGenerator.(*MockSourceMapGenerator)
			if !ok {
				t.Fatal("Expected MockSourceMapGenerator")
			}
			if content, exists := mockGen.SourceContents["test.less"]; !exists || content != "body { color: red; }" {
				t.Errorf("Source content should have ignored chars removed: got %v", content)
			}
		})

		t.Run("should set sourceMapURL from options", func(t *testing.T) {
			sourceMapOutput.sourceMapURL = "custom.map"
			sourceMapOutput.rootNode = &MockNode{
				GenCSSFunc: func(context map[string]any, output *SourceMapOutput) {
					output.Add("body { color: red; }", nil, 0, false)
				},
			}

			sourceMapOutput.ToCSS(mockContext)
			
			if sourceMapOutput.sourceMapURL != "custom.map" {
				t.Errorf("sourceMapURL = %v, want %v", sourceMapOutput.sourceMapURL, "custom.map")
			}
		})

		t.Run("should use sourceMapFilename as URL when no sourceMapURL", func(t *testing.T) {
			sourceMapOutput.sourceMapFilename = "output.css.map"
			sourceMapOutput.sourceMapURL = ""
			sourceMapOutput.rootNode = &MockNode{
				GenCSSFunc: func(context map[string]any, output *SourceMapOutput) {
					output.Add("body { color: red; }", nil, 0, false)
				},
			}

			sourceMapOutput.ToCSS(mockContext)
			
			if sourceMapOutput.sourceMapURL != "output.css.map" {
				t.Errorf("sourceMapURL = %v, want %v", sourceMapOutput.sourceMapURL, "output.css.map")
			}
		})

		t.Run("should return empty string when no CSS generated", func(t *testing.T) {
			sourceMapOutput := NewSourceMapOutput(options)
			css := sourceMapOutput.ToCSS(mockContext)
			if css != "" {
				t.Errorf("ToCSS() = %v, want empty string", css)
			}
			if sourceMapOutput.SourceMap != "" {
				t.Error("SourceMap should be empty when no CSS generated")
			}
		})

		t.Run("should not set source content when outputSourceFiles is false", func(t *testing.T) {
			options := SourceMapOutputOptions{
				RootNode: &MockNode{},
				ContentsMap: map[string]string{
					"test.less": "body { color: red; }",
				},
				ContentsIgnoredCharsMap: map[string]int{},
				OutputSourceFiles:      false,
				SourceMapGeneratorConstructor: createMockSourceMapGenerator,
			}
			sourceMapOutput := NewSourceMapOutput(options)
			sourceMapOutput.rootNode = &MockNode{
				GenCSSFunc: func(context map[string]any, output *SourceMapOutput) {
					output.Add("body { color: red; }", nil, 0, false)
				},
			}

			sourceMapOutput.ToCSS(mockContext)
			
			mockGen, ok := sourceMapOutput.sourceMapGenerator.(*MockSourceMapGenerator)
			if !ok {
				t.Fatal("Expected MockSourceMapGenerator")
			}
			if len(mockGen.SourceContents) != 0 {
				t.Error("Source content should not be set when outputSourceFiles is false")
			}
		})
	})
}

File: packages/less/src/less/source-map-output.test.js
import { describe, it, expect, beforeEach, vi } from 'vitest';
import SourceMapOutputFactory from './source-map-output.js';

describe('SourceMapOutput', () => {
    let mockEnvironment;
    let SourceMapOutput;
    let mockSourceMapGenerator;
    let mockRootNode;

    beforeEach(() => {
        // Mock source map generator
        mockSourceMapGenerator = {
            addMapping: vi.fn(),
            setSourceContent: vi.fn(),
            toJSON: vi.fn(() => ({ version: 3, sources: ['test.less'], mappings: 'AAAA' }))
        };

        // Mock environment
        mockEnvironment = {
            getSourceMapGenerator: vi.fn(() => function MockSourceMapGenerator() {
                return mockSourceMapGenerator;
            }),
            encodeBase64: vi.fn((str) => Buffer.from(str).toString('base64'))
        };

        // Mock root node
        mockRootNode = {
            genCSS: vi.fn()
        };

        SourceMapOutput = SourceMapOutputFactory(mockEnvironment);
    });

    describe('constructor', () => {
        it('should initialize with basic options', () => {
            const options = {
                rootNode: mockRootNode,
                contentsMap: { 'test.less': 'body { color: red; }' },
                contentsIgnoredCharsMap: {},
                outputFilename: 'output.css'
            };

            const sourceMapOutput = new SourceMapOutput(options);
            
            expect(sourceMapOutput._css).toEqual([]);
            expect(sourceMapOutput._rootNode).toBe(mockRootNode);
            expect(sourceMapOutput._contentsMap).toBe(options.contentsMap);
            expect(sourceMapOutput._outputFilename).toBe('output.css');
            expect(sourceMapOutput._lineNumber).toBe(0);
            expect(sourceMapOutput._column).toBe(0);
        });

        it('should normalize Windows paths in sourceMapFilename', () => {
            const options = {
                rootNode: mockRootNode,
                contentsMap: {},
                contentsIgnoredCharsMap: {},
                sourceMapFilename: 'C:\\path\\to\\map.css.map'
            };

            const sourceMapOutput = new SourceMapOutput(options);
            expect(sourceMapOutput._sourceMapFilename).toBe('C:/path/to/map.css.map');
        });

        it('should normalize and add trailing slash to sourceMapRootpath', () => {
            const options = {
                rootNode: mockRootNode,
                contentsMap: {},
                contentsIgnoredCharsMap: {},
                sourceMapRootpath: 'C:\\src\\styles'
            };

            const sourceMapOutput = new SourceMapOutput(options);
            expect(sourceMapOutput._sourceMapRootpath).toBe('C:/src/styles/');
        });

        it('should handle sourceMapRootpath that already has trailing slash', () => {
            const options = {
                rootNode: mockRootNode,
                contentsMap: {},
                contentsIgnoredCharsMap: {},
                sourceMapRootpath: '/src/styles/'
            };

            const sourceMapOutput = new SourceMapOutput(options);
            expect(sourceMapOutput._sourceMapRootpath).toBe('/src/styles/');
        });

        it('should set empty sourceMapRootpath when not provided', () => {
            const options = {
                rootNode: mockRootNode,
                contentsMap: {},
                contentsIgnoredCharsMap: {}
            };

            const sourceMapOutput = new SourceMapOutput(options);
            expect(sourceMapOutput._sourceMapRootpath).toBe('');
        });

        it('should normalize Windows paths in sourceMapBasepath', () => {
            const options = {
                rootNode: mockRootNode,
                contentsMap: {},
                contentsIgnoredCharsMap: {},
                sourceMapBasepath: 'C:\\project\\src'
            };

            const sourceMapOutput = new SourceMapOutput(options);
            expect(sourceMapOutput._sourceMapBasepath).toBe('C:/project/src');
        });
    });

    describe('removeBasepath', () => {
        let sourceMapOutput;

        beforeEach(() => {
            const options = {
                rootNode: mockRootNode,
                contentsMap: {},
                contentsIgnoredCharsMap: {},
                sourceMapBasepath: '/project/src'
            };
            sourceMapOutput = new SourceMapOutput(options);
        });

        it('should remove basepath when path starts with basepath', () => {
            const result = sourceMapOutput.removeBasepath('/project/src/styles/main.less');
            expect(result).toBe('styles/main.less');
        });

        it('should remove basepath and leading slash/backslash', () => {
            sourceMapOutput._sourceMapBasepath = '/project/src';
            expect(sourceMapOutput.removeBasepath('/project/src/main.less')).toBe('main.less');
            expect(sourceMapOutput.removeBasepath('/project/src\\main.less')).toBe('main.less');
        });

        it('should return original path when basepath does not match', () => {
            const result = sourceMapOutput.removeBasepath('/other/path/main.less');
            expect(result).toBe('/other/path/main.less');
        });

        it('should handle empty basepath', () => {
            sourceMapOutput._sourceMapBasepath = '';
            const result = sourceMapOutput.removeBasepath('/path/to/file.less');
            expect(result).toBe('/path/to/file.less');
        });
    });

    describe('normalizeFilename', () => {
        let sourceMapOutput;

        beforeEach(() => {
            const options = {
                rootNode: mockRootNode,
                contentsMap: {},
                contentsIgnoredCharsMap: {},
                sourceMapBasepath: '/project/src',
                sourceMapRootpath: '/assets/'
            };
            sourceMapOutput = new SourceMapOutput(options);
        });

        it('should normalize Windows paths to forward slashes', () => {
            const result = sourceMapOutput.normalizeFilename('C:\\project\\src\\styles\\main.less');
            expect(result).toBe('/assets/C:/project/src/styles/main.less');
        });

        it('should combine rootpath with filename after removing basepath', () => {
            const result = sourceMapOutput.normalizeFilename('/project/src/main.less');
            expect(result).toBe('/assets/main.less');
        });

        it('should handle empty rootpath', () => {
            sourceMapOutput._sourceMapRootpath = '';
            const result = sourceMapOutput.normalizeFilename('/project/src/main.less');
            expect(result).toBe('main.less');
        });
    });

    describe('add', () => {
        let sourceMapOutput;

        beforeEach(() => {
            const options = {
                rootNode: mockRootNode,
                contentsMap: {
                    'test.less': 'body {\n  color: red;\n}'
                },
                contentsIgnoredCharsMap: {}
            };
            sourceMapOutput = new SourceMapOutput(options);
        });

        it('should ignore empty chunks', () => {
            sourceMapOutput.add('');
            sourceMapOutput.add(null);
            sourceMapOutput.add(undefined);
            
            expect(sourceMapOutput._css).toEqual([]);
        });

        it('should add chunk without file info', () => {
            sourceMapOutput.add('body { color: blue; }');
            
            expect(sourceMapOutput._css).toEqual(['body { color: blue; }']);
            expect(sourceMapOutput._column).toBe(21);
            expect(sourceMapOutput._lineNumber).toBe(0);
        });

        it('should handle multi-line chunks', () => {
            sourceMapOutput.add('body {\n  color: blue;\n}');
            
            expect(sourceMapOutput._lineNumber).toBe(2);
            expect(sourceMapOutput._column).toBe(1);
        });

        it('should add chunk with file info and create mapping', () => {
            sourceMapOutput._sourceMapGenerator = mockSourceMapGenerator;
            const fileInfo = { filename: 'test.less' };
            
            sourceMapOutput.add('body { color: blue; }', fileInfo, 5);
            
            expect(mockSourceMapGenerator.addMapping).toHaveBeenCalledWith({
                generated: { line: 1, column: 0 },
                original: { line: 1, column: 5 },
                source: 'test.less'
            });
        });

        it('should handle mapLines option', () => {
            sourceMapOutput._sourceMapGenerator = mockSourceMapGenerator;
            const fileInfo = { filename: 'test.less' };
            
            sourceMapOutput.add('line1\nline2\nline3', fileInfo, 0, true);
            
            expect(mockSourceMapGenerator.addMapping).toHaveBeenCalledTimes(3);
            expect(mockSourceMapGenerator.addMapping).toHaveBeenNthCalledWith(1, {
                generated: { line: 1, column: 0 },
                original: { line: 1, column: 0 },
                source: 'test.less'
            });
            expect(mockSourceMapGenerator.addMapping).toHaveBeenNthCalledWith(2, {
                generated: { line: 2, column: 0 },
                original: { line: 2, column: 0 },
                source: 'test.less'
            });
        });

        it('should handle ignored characters', () => {
            sourceMapOutput._contentsIgnoredCharsMap = { 'test.less': 5 };
            sourceMapOutput._sourceMapGenerator = mockSourceMapGenerator;
            const fileInfo = { filename: 'test.less' };
            
            sourceMapOutput.add('color: blue;', fileInfo, 10);
            
            expect(mockSourceMapGenerator.addMapping).toHaveBeenCalledWith({
                generated: { line: 1, column: 0 },
                original: { line: 2, column: 3 },
                source: 'test.less'
            });
        });

        it('should handle negative index after ignored chars adjustment', () => {
            sourceMapOutput._contentsIgnoredCharsMap = { 'test.less': 10 };
            sourceMapOutput._sourceMapGenerator = mockSourceMapGenerator;
            const fileInfo = { filename: 'test.less' };
            
            sourceMapOutput.add('color: blue;', fileInfo, 5);
            
            expect(mockSourceMapGenerator.addMapping).toHaveBeenCalledWith({
                generated: { line: 1, column: 0 },
                original: { line: 1, column: 0 },
                source: 'test.less'
            });
        });

        it('should handle missing content in contentsMap', () => {
            const fileInfo = { filename: 'missing.less' };
            
            sourceMapOutput.add('body { color: blue; }', fileInfo, 0);
            
            expect(sourceMapOutput._css).toEqual(['body { color: blue; }']);
        });
    });

    describe('isEmpty', () => {
        let sourceMapOutput;

        beforeEach(() => {
            const options = {
                rootNode: mockRootNode,
                contentsMap: {},
                contentsIgnoredCharsMap: {}
            };
            sourceMapOutput = new SourceMapOutput(options);
        });

        it('should return true when no CSS has been added', () => {
            expect(sourceMapOutput.isEmpty()).toBe(true);
        });

        it('should return false when CSS has been added', () => {
            sourceMapOutput.add('body { color: red; }');
            expect(sourceMapOutput.isEmpty()).toBe(false);
        });
    });

    describe('toCSS', () => {
        let sourceMapOutput;
        let mockContext;

        beforeEach(() => {
            mockContext = {};
            const options = {
                rootNode: mockRootNode,
                contentsMap: {
                    'test.less': 'body { color: red; }'
                },
                contentsIgnoredCharsMap: {},
                outputFilename: 'output.css',
                outputSourceFiles: true
            };
            sourceMapOutput = new SourceMapOutput(options);
        });

        it('should generate CSS and source map', () => {
            mockRootNode.genCSS.mockImplementation((context, output) => {
                output.add('body { color: red; }');
            });

            const css = sourceMapOutput.toCSS(mockContext);
            
            expect(mockRootNode.genCSS).toHaveBeenCalledWith(mockContext, sourceMapOutput);
            expect(css).toBe('body { color: red; }');
            expect(sourceMapOutput.sourceMap).toBeDefined();
        });

        it('should set source content when outputSourceFiles is true', () => {
            mockRootNode.genCSS.mockImplementation((context, output) => {
                output.add('body { color: red; }');
            });

            sourceMapOutput.toCSS(mockContext);
            
            expect(mockSourceMapGenerator.setSourceContent).toHaveBeenCalledWith(
                'test.less',
                'body { color: red; }'
            );
        });

        it('should handle ignored characters in source content', () => {
            sourceMapOutput._contentsIgnoredCharsMap = { 'test.less': 5 };
            sourceMapOutput._contentsMap = { 'test.less': '/* */body { color: red; }' };
            
            mockRootNode.genCSS.mockImplementation((context, output) => {
                output.add('body { color: red; }');
            });

            sourceMapOutput.toCSS(mockContext);
            
            expect(mockSourceMapGenerator.setSourceContent).toHaveBeenCalledWith(
                'test.less',
                'body { color: red; }'
            );
        });

        it('should set sourceMapURL from options', () => {
            sourceMapOutput.sourceMapURL = 'custom.map';
            
            mockRootNode.genCSS.mockImplementation((context, output) => {
                output.add('body { color: red; }');
            });

            sourceMapOutput.toCSS(mockContext);
            
            expect(sourceMapOutput.sourceMapURL).toBe('custom.map');
        });

        it('should use sourceMapFilename as URL when no sourceMapURL', () => {
            sourceMapOutput._sourceMapFilename = 'output.css.map';
            
            mockRootNode.genCSS.mockImplementation((context, output) => {
                output.add('body { color: red; }');
            });

            sourceMapOutput.toCSS(mockContext);
            
            expect(sourceMapOutput.sourceMapURL).toBe('output.css.map');
        });

        it('should return empty string when no CSS generated', () => {
            const css = sourceMapOutput.toCSS(mockContext);
            expect(css).toBe('');
            expect(sourceMapOutput.sourceMap).toBeUndefined();
        });

        it('should not set source content when outputSourceFiles is false', () => {
            sourceMapOutput._outputSourceFiles = false;
            
            mockRootNode.genCSS.mockImplementation((context, output) => {
                output.add('body { color: red; }');
            });

            sourceMapOutput.toCSS(mockContext);
            
            expect(mockSourceMapGenerator.setSourceContent).not.toHaveBeenCalled();
        });
    });
});